src/adrs/0129349218/README.md

```diff
@@ -0,0 +1,532 @@
+you may need to merge the `subject` when wrapping errors.
+
+
+## Pending Concerns
+
+- Having infra-level `info` where you could use the `metadata` to identify the
+  context in which the error happened introduces a level of complexity that
+  it is desired to be avoided. The `domain` and `type` fields should be enough
+  to identify the context in which the error happened. Example:
+  `googleapis.com` domain and `NOT_FOUND` reason. Most likely the `metadata`
+  field will be used to provide more context to the error message, example:
+  `resource_type`. If `id` must be unique across the entire domain, then
+  it becomes a coordination problem, and **it is desired to avoid it**.
+- Message placeholder comes from C# ecosystem, what is not desired here is to
+  introduce security vulnerabilities, look at what happened to Java log4j
+  vulnerability. So, is this desired?
+- Should Message Placeholder allow to replace with the `info` `metadata`?
+  Probably not, since whoever defines the `info` and its structure owns the
+  data, and the structure of the error
+- `source?: Source;` I am not sure about the naming.
+- Under `info.data`
+
+## Resolution
+
+- You **MUST** use the `application/universal-error+json` meme when sending the
+  error over the wire as a JSON payload.
+
+```ts
+type UniversalError = {
+  /**
+   * The version of the specification that the error uses.
+   */
+  specversion: string;
+  code: Code;
+  message: MessageDescription;
+  info: ErrorInfo;
+  causes: UniversalError[];
+
+  //
+  // Optional Information
+  //
+  source?: Source;
+  indeterministic_info?: IndeterministicInfo;
+  help?: Help;
+  debug_info?: DebugInfo;
+  localized_message?: LocalizedMessage;
+  retry_info?: RetryInfo;
+};
+
+enum Code {
+  CANCELLED,
+  UNKNOWN,
+  INVALID_ARGUMENT,
+  DEADLINE_EXCEEDED,
+  NOT_FOUND,
+  ALREADY_EXISTS,
+  PERMISSION_DENIED,
+  UNAUTHENTICATED,
+  RESOURCE_EXHAUSTED,
+  FAILED_PRECONDITION,
+  ABORTED,
+  OUT_OF_RANGE,
+  UNIMPLEMENTED,
+  INTERNAL,
+  UNAVAILABLE,
+  DATA_LOSS
+}
+
+/**
+ * Message contains a generic description of the error condition in English.
+ *
+ * Warning: Error messages are not part of the API surface. They are subject to
+ * changes without notice. Application code must not have a hard dependency on
+ * error messages.
+ *
+ * It is intended for a human audience. Simple programs display the message
+ * directly to the end user if they encounter an error condition they don't
+ * know how or don't care to handle. Sophisticated programs with more
+ * exhaustive error handling and proper internationalization are more likely
+ * to ignore the error message.
+ *
+ * Messages MAY use {placeholders} to reference values provided in the
+ * `info.metadata` field. For example, "The {account_id} is not found.".
+ */
+type MessageDescription = string;
+
+
+/**
+ * Source is the source of the error event. It contains information about the
+ * producer of the error event.
+ */
+type Source = {
```

Comment:
Should sources be exposed to end users? What if they refer to "meaningless" internal microservices?

I can imagine sources being translated before end user exposure

Reply:
> Should sources be exposed to end users?
Yes

> What if they refer to "meaningless" internal microservices?
Then, the clients decide to ignore the source, and there is not much to do on their end.

In practice, you wouldn't have that many of those errors that are not related to a subject from a given Command Message.